### ISSUE TITLE
AMD MI100 only got 10% performance of Nvidia A100 of MultiRHSProf 

### ISSUE THREAD

**Issue by Alice1069** [LatticeQCD/SIMULATeQCD#82]
AMD MI100 got only 10% performance of Nvidia A100, am I doing any configuration wrong, or it's expected result? 
as MI100 has similar Tops/memory bw/power with A100, i expect performance should also be similar.

AMD MI100 performance:
[2022-05-02 09:37:52] INFO: Starting Test with 1 Stacks
...
[2022-05-02 09:37:54] INFO: Achieved TFLOP/s **0.10231**9

Nvidia A100 performance:
[2022-05-02 09:47:34] INFO: Starting Test with 1 Stacks
...
[2022-05-02 09:47:34] INFO: Achieved TFLOP/s **0.971428**

----------------------------------------------------------
---reproduce steps:-----------------

-------common setting-------------
AMD and Nvidia has same code, git commit is: 
commit 01a0c7170cae7eec4652ee6c24af903a8a676725 (HEAD -> main, origin/main, origin/HEAD)
Author: lukas-mazur <lukas.mazur@uni-paderborn.de>
Date:   Sun May 1 10:14:41 2022 +0200

    hip_backend ARCHITECTURE fix in communicationBase_mpi.cpp
-------common setting 2------------
after compile change build_folder/parameter/profiling/MrhsDSlashProf.param
Nodes       = 2 1 1 1 -> 1 1 1 1 //so to use only 1 gpu

------AMD MI100 steps-----------

export HIP_PATH=/opt/rocm/hip
export CC=/opt/rocm/llvm/bin/clang
export CXX=/opt/rocm/llvm/bin/clang++

CXX=hipcc \
cmake ../SIMULATeQCD/ \
-DARCHITECTURE="gfx908" \
-DUSE_GPU_AWARE_MPI=ON \
-DUSE_GPU_P2P=OFF \
-DBACKEND="hip_amd" \
-DMPI_CXX_COMPILER=/home/gputest/alice/openmpi/bin/mpicxx \
-DMPI_C_COMPILER=/home/gputest/alice/openmpi/bin/mpicc

-- Using HIP backend for AMD GPUs. (Experimental!)
-- hip::amdhip64 is SHARED_LIBRARY
-- hip::amdhip64 is SHARED_LIBRARY
-- HIP (AMD) Architecture: gfx908
-- Configuring done
-- Generating done
-- Build files have been written to: /home/gputest/alice/SIMULATeQCD_build2

make MultiRHSProf -j 60

gputest@gpuserver:~/alice/SIMULATeQCD_build2$ mpiexec -np 1 ./MultiRHSProf
[2022-05-02 09:37:51] INFO: Initializing MPI with (1 proc)
[2022-05-02 09:37:51] INFO: Reading parameters from file :: ../parameter/profiling/MrhsDSlashProf.param
[2022-05-02 09:37:51] INFO: # PARAM :: Lattice = 20 20 20 20
[2022-05-02 09:37:51] INFO: # PARAM :: Nodes = 1 1 1 1
[2022-05-02 09:37:51] INFO: # PARAM :: Topology = 0 0 0 0  (default)
[2022-05-02 09:37:51] INFO: # PARAM :: endianness = auto (default)
[2022-05-02 09:37:51] INFO: # PARAM :: prec_out = 0 (default)
[2022-05-02 09:37:51] INFO: # PARAM :: measurements_dir = ./ (default)
[2022-05-02 09:37:51] INFO: # PARAM :: use_unit_conf = 0 (default)
[2022-05-02 09:37:51] INFO: > Running on:
[2022-05-02 09:37:51] INFO: [Rank 0] > gpuserver CPU_0 GPU_3D; MPI: world_rank=0, node_rank=0, coord: ( 0 0 0 0 )
[2022-05-02 09:37:51] INFO: > Is MPI CUDA-aware? @compilation NO, @runtime NO
[2022-05-02 09:37:51] INFO: > GPU compute capability: 90
[2022-05-02 09:37:51] WARN: Cannot determine for which compute capability the code was compiled!
[2022-05-02 09:37:51] INFO: > Checking support for P2P (Peer-to-Peer) and UVA (Unified Virtual Addressing):
[2022-05-02 09:37:51] INFO: [Rank 0] > gpuserver GPU_3D(): P2P NO ; UVA Unknown (HIP does not support this!)
[2022-05-02 09:37:51] INFO: Initialize Indexer with HaloDepth = 4
[2022-05-02 09:37:52] INFO: -------------------------------------
[2022-05-02 09:37:52] INFO: Running on Device
[2022-05-02 09:37:52] INFO: -------------------------------------
[2022-05-02 09:37:52] INFO: Testing Even - Odd
[2022-05-02 09:37:52] INFO: ------------------
[2022-05-02 09:37:52] INFO: Starting Test with 1 Stacks
[2022-05-02 09:37:52] INFO: Initialize random state
[2022-05-02 09:37:52] INFO: gen conf
[2022-05-02 09:37:52] INFO: Initialize spinors
[2022-05-02 09:37:52] INFO: Randomize spinors
[2022-05-02 09:37:52] INFO: Initialize DSlash
[2022-05-02 09:37:54] INFO: Time for 500 applications of multiRHS Dslash: 905.400ms
[2022-05-02 09:37:54] INFO: Achieved TFLOP/s 0.102319

------Nvidia A100 steps-----------
using nvidia docker container tensorrt-20.07(cuda11.0)

cmake ../SIMULATeQCD/ \
-DARCHITECTURE="80" \
-DUSE_GPU_AWARE_MPI=ON \
-DUSE_GPU_P2P=ON \
 
make MultiRHSProf -j 60

mpiexec -np 1 --allow-run-as-root ./MultiRHSProf
--------------------------------------------------------------------------
[[28245,1],0]: A high-performance Open MPI point-to-point messaging module
was unable to find any relevant network interfaces:

Module: OpenFabrics (openib)
  Host: 9d88439594e4

Another transport will be used instead, although this may result in
lower performance.

NOTE: You can disable this warning by setting the MCA parameter
btl_base_warn_component_unused to 0.
--------------------------------------------------------------------------
[2022-05-02 09:47:34] INFO: Initializing MPI with (1 proc)
[2022-05-02 09:47:34] INFO: Reading parameters from file :: ../parameter/profiling/MrhsDSlashProf.param
[2022-05-02 09:47:34] INFO: # PARAM :: Lattice = 20 20 20 20
[2022-05-02 09:47:34] INFO: # PARAM :: Nodes = 1 1 1 1
[2022-05-02 09:47:34] INFO: # PARAM :: Topology = 0 0 0 0  (default)
[2022-05-02 09:47:34] INFO: # PARAM :: endianness = auto (default)
[2022-05-02 09:47:34] INFO: # PARAM :: prec_out = 0 (default)
[2022-05-02 09:47:34] INFO: # PARAM :: measurements_dir = ./ (default)
[2022-05-02 09:47:34] INFO: # PARAM :: use_unit_conf = 0 (default)
[2022-05-02 09:47:34] INFO: > Running on:
[2022-05-02 09:47:34] INFO: [Rank 0] > 9d88439594e4 CPU_0 GPU_1; MPI: world_rank=0, node_rank=0, coord: ( 0 0 0 0 )
[2022-05-02 09:47:34] INFO: > Is MPI CUDA-aware? @compilation NO, @runtime NO
[2022-05-02 09:47:34] INFO: > GPU compute capability: 80
[2022-05-02 09:47:34] INFO: > Checking support for P2P (Peer-to-Peer) and UVA (Unified Virtual Addressing):
[2022-05-02 09:47:34] INFO: [Rank 0] > 9d88439594e4 GPU_1(NVIDIA A100-PCIE-40GB): P2P YES; UVA YES
[2022-05-02 09:47:34] INFO: Initialize Indexer with HaloDepth = 4
[2022-05-02 09:47:34] INFO: -------------------------------------
[2022-05-02 09:47:34] INFO: Running on Device
[2022-05-02 09:47:34] INFO: -------------------------------------
[2022-05-02 09:47:34] INFO: Testing Even - Odd
[2022-05-02 09:47:34] INFO: ------------------
[2022-05-02 09:47:34] INFO: Starting Test with 1 Stacks
[2022-05-02 09:47:34] INFO: Initialize random state
[2022-05-02 09:47:34] INFO: gen conf
[2022-05-02 09:47:34] INFO: Initialize spinors
[2022-05-02 09:47:34] INFO: Randomize spinors
[2022-05-02 09:47:34] INFO: Initialize DSlash
[2022-05-02 09:47:34] INFO: Time for 500 applications of multiRHS Dslash: 95.365ms
[2022-05-02 09:47:34] INFO: Achieved TFLOP/s 0.971428

please let me know if more information needs to be provided.

**Comment by lukas-mazur** [comment]
@philomat did some tests on an AMD MI100 GPU as well. And as far as I know he also saw that the performance got worse on this device, because:
- the Vector ALU is only busy 20-30% of the time (depending on the # RHS)
- the memory unit is busy about 90% of the time but that includes 20% (of the total GPU time) of stalling.

Maybe @philomat can comment on that in a bit more detail.

Unfortunately, I can't investigate this by myself, as I don't have access to an AMD GPU.
However, keep in mind that the HIP backend is still in experimental mode. So we do not recommend running production jobs with this backend.

**Comment by philomat** [comment]
Hi Alice,

I suggest you increase the lattice size. I actually have no idea, why our default parameter is still 20^4. This is way to small to saturate the GPU. 

Also, try to remove the #pragma unroll in l. 16 of src/modules/dslash/dslash.cpp. This pragma gives better performance on A100 but reduces performance on AMD GPUs (I don't think is was much but I do not remember anymore).

I used 64^3 x 20 in my profiling runs. With this lattice size I get: 

- on A100 with CUDA 11.0:  1.33 TFLOP/s
- on A100 with CUDA >= 11.3: 1.28 TFLOP/s

- on MI100 with ROCm 4.5: 0.69 TFLOP/s

So basically I performance on MI100 is 0.5* performance on A100. And I do not see any scaling with the number of RHS on MI100.

@lukas-mazur is right, the Vector ALU is only busy ~20% of the time and memory unit ~80%. 

I think the lack of performance on AMD systems (numbers for MI250 are similar) is connected to the fact that our code seems to use a large amount of scalar register. And this register pressure seem to lead to a low occupancy. 
Unfortunately, the rocm profiler does not come close to the products by NVIDIA or Intel, so this is more of an educated guess than proper analysis.

I have not seen benchmarks or profiles from other codes but I remember having a conversation with Peter Boyle about Grid and he said that one Crusher node (4 MI250 Cards, i.e. 8 MI100 like GPUs) is comparable to one Perlmutter node (4 A100 GPUs), performance wise. So very similar to our case.

What is somewhat puzzling me, is that we do not see any scaling with the number of RHS on AMD. This might be related to the small amount of L2 cache. But on the other hand V100 also has rather small cache sizes and here we find scaling with #RHS.

**Comment by Alice1069** [comment]
with Lattice = 64 64 64 20

A100-pcie-40GB cuda11.0
 1rhs: Achieved TFLOP/s 1.28358 (TFLOP/s 0.97 at Lattice 20 20 20 20)
10rhs: Achieved TFLOP/s 2.464   (max performance)
12rhs: Achieved TFLOP/s 2.33663

AMD MI100 
 1rhs: Achieved TFLOP/s 0.189207  (TFLOP/s 0.10 at Lattice 20 20 20 20)
12rhs: Achieved TFLOP/s 0.176722

a glimpse of rocm-smi:
GPU  Temp   AvgPwr  SCLK     MCLK     Fan  Perf    PwrCap  VRAM%  GPU%
0    63.0c  285.0W  1502Mhz  1200Mhz  0%   manual  290.0W   24%   99%

----------after doing below modification as you recommended:--------------
--- a/src/modules/dslash/dslash.cpp
+++ b/src/modules/dslash/dslash.cpp
@@ -13,7 +13,7 @@ __host__ __device__ auto HisqDslashFunctor<floatT, LatLayoutRHS, HaloDepthGauge,
     typedef GIndexer<LayoutSwitcher<LatLayoutRHS>(), HaloDepthSpin> GInd;

     gVect3<floatT> Stmp(0.0);
-#pragma unroll
+//#pragma unroll
     for (int mu = 0; mu < 4; mu++) {
-------------------------------------------------------
 1rhs: Achieved TFLOP/s 0.788867 (ROCm 5.0)
 4rhs: Achieved TFLOP/s 1.12238 (max performance)
12rhs: Achieved TFLOP/s 1.05022

a glimpse of rocm-smi:
GPU  Temp   AvgPwr  SCLK     MCLK     Fan  Perf    PwrCap  VRAM%  GPU%
0    80.0c  293.0W  1472Mhz  1200Mhz  0%   manual  290.0W   32%   99%

some points:
1. thanks for your patch, MI100 1rhs performance increased from 0.189 to 0.788, (while VRAM does not change much 25%~27%)
2. MI100's has higher performance on 4/8/12 rhs.
3. MI100's max rhs scale is 1.42x, A100's max scale is 2.46x, as you said AMD has bad rhs scaling.
4. now MI100's 1rhs performance 0.788 is 61% of A100 1.283. 
look at the MI100/A100 specification:
	fp32-Tflop	HBM_BW-GB/s	power	freq-Mhz
MI100	23.1/46.1	1229		290w	1502
A100	19.5		1555		250w	1410
consider 1rhs should be a HBM bottleneck workload, HBM bw ratio is 1229/1555=79%. consider MI100 has lower stream efficiency, the final HBM bw ratio is 74%.
so the gap between 1rhs performance 61% to HBM bw ratio 74% is 13%, not sure if 13% is explainable with scale register usage or any chance to improve that?

Finally, thanks again for the patch.

**Comment by philomat** [comment]
Hi again,

I have another patch that further improves performance on AMD GPUs. It basically reduces register pressure by using a simpler indexing scheme. If I remember correctly the improvement was about 10%. I have not pushed it because it is only beneficial on AMD, on NVIDIA is actually detrimental and costs performance. Further, I have not tested it for correctness. 

If you are interested in that, check out the test_better_dslash branch.

**Comment by Alice1069** [comment]
test report: I switched to test_better_dslash branch. since build error, i removed libmpi_gtl_hsa.so line:
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,7 +75,7 @@ if (USE_HIP)
     endif()

     set(SIMULATEQCD_INCLUDE_DIRS "${SIMULATEQCD_INCLUDE_DIRS};${HIP_PATH}/include;${HIP_INCLUDE_DIRS}")
-    set(SIMULATEQCD_LIBRARIES "${SIMULATEQCD_LIBRARIES};${HIP_LIBRARIES};/opt/cray/pe/mpich/8.1.12/gtl/lib/libmpi_gtl_hsa.so")
+    set(SIMULATEQCD_LIBRARIES "${SIMULATEQCD_LIBRARIES};${HIP_LIBRARIES}")  

 endif()

I make and run MultiRHSProf on MI100.

Lattice = 72 72 72 8 
1rhs: Achieved TFLOP/s 0.739301

Lattice = 64 64 64 20
1rhs: Achieved TFLOP/s 0.791832 (compare to main branch + "//#pragma unroll" patch: 0.788867,almost same. maybe rocm5 already digest that improvement. or maybe, it will improve on MI250, not for MI100.)
4rhs: Achieved TFLOP/s 1.12762 (compare to main branch + "//#pragma unroll" patch: 1.12238, almost same )

Thanks for your information. I'd like to be informed when you have further update.
I am ok to close this bug if you have project management consideration.

**Comment by Alice1069** [comment]
@philomat would you like to share your MI250 performance number on test_better_dslash branch? I could double check that after I get MI250 hardware.

**Comment by philomat** [comment]
@Alice1069 I will have to get permission to share performance numbers for MI250. Let me see what I can do

**Comment by Alice1069** [comment]
@philomat hi, I just test latest code on MI210, I only got 0.14 TOPS, 
if i removed #program unroll，I got 0.18TOPS. 
compares to previous MI100 0.78 Tops, it's pretty low. 
I am not sure is there any method could improve this?

Lattice = 64 64 64 20
SIMULATeQCD code: main branch, git commit 2022 Aug 2, 551ccca0
-DARCHITECTURE="gfx90a"  //for mi210
mpiexec -np 1 --allow-run-as-root ./MultiRHSProf (1rhs)

I am wondering whether it because the rocm version changed from 5.0 to 5.2 or SIMULATeQCD's new patches from May to Aug?
